### PR TITLE
Docker image workflow (publish on `ghcr.io`)

### PR DESCRIPTION
**Pull request by AFCMS**
Should fix https://github.com/minetest/minetest/issues/14164

Follows the official Docker documentation to build image using GitHub Actions workflows.

Official Actions by Docker are used.

Links:
- https://docs.github.com/en/actions/publishing-packages/publishing-docker-images
- https://docs.docker.com/build/ci/github-actions/multi-platform
- https://github.com/opencontainers/image-spec/blob/main/annotations.md

Tags are automatically determined by the Docker metadata Action.
You can find how its done [here](https://github.com/docker/metadata-action?tab=readme-ov-file#basic)

~~**Please note that the build of the `arm64` image takes a much longer time than the `amd64` one. This is due to the use of QEMU for the build process, following the Docker documentation. It may be possible to not require the use of QEMU by modifying the Dockerfile but I don't have the knowledge to do so.**~~

## To do

This PR is a Work in Progress.

- [x] Basic stuff
- [x] Test if build works (on fork)
- [x] Is opencontainers metadata correct for the images?
- [x] ~~Should the workflow use a matrix strategy for the architectures?~~ No more arm64
- [x] Remove old image building workflow?
- [x] Test running `--version` in CI
- [x] ~~More architectures? ([alpine supported architectures](https://hub.docker.com/_/alpine/tags)) Or remove `arm64`?~~ Require cross-compilation support in Dockerfile.

## How to test

Merge the branch into one of your forks of Minetest, see the workflow run, see the container in the package section of the repository. See the two architectures. Pull the image and see if it runs.

In my case:
- https://github.com/AFCMS/minetest/actions/runs/7448427624/job/20262783153
- https://github.com/AFCMS/minetest/pkgs/container/minetest
- `docker run ghcr.io/afcms/minetest:master /usr/local/bin/minetestserver --version` works

### CODE REVIEW

**Comment by sfan5** [comment]
> Please note that the build of the `arm64` image takes a much longer time than the `amd64` one. This is due to the use of QEMU for the build process, following the Docker documentation. It may be possible to not require the use of QEMU by modifying the Dockerfile but I don't have the knowledge to do so.

The use of qemu cannot be avoided since Github does not provide any ARM runners. I'd simply drop it.

**Comment by AFCMS** [comment]
> The use of qemu cannot be avoided since Github does not provide any ARM runners. I'd simply drop it.

Isn't arm support interesting enough for raspberry pi users? The CI is basically free so my biggest concern was the longer time for the CI to finish in stuff like PRs. That could perhaps be improved by using a matrix strategy so the two builds are visually separated.

**Comment by AFCMS** [comment]
Docker support cross compilation in the image build process, but it requires modifications of the Dockerfile.

https://docs.docker.com/build/guide/multi-platform/#build-using-cross-compilation

GCC is also not straightforward to setup for cross-compilation, so I think its better to just drop the QEMU based arm build and later think about allowing cross-compilation in the the Dockerfile.

**Comment by AFCMS** [comment]
I dropped the `arm64` support. IMO it would be a good thing to support it, but it should be done with cross-compilation support in the Dockerfile.

**Comment by AFCMS** [comment]
On the topic of cross compilation, this seems to be a pretty useful tool:

https://github.com/tonistiigi/xx

File: .github/workflows/docker_image.yml
Comment (sfan5) [inline .github/workflows/docker_image.yml:9]: can this match against paths too like existing workflows?

File: .github/workflows/docker_image.yml
Comment (sfan5) [inline .github/workflows/docker_image.yml:15]: this duplicates what the existing workflow does: https://github.com/minetest/minetest/blob/master/.github/workflows/linux.yml#L155

We should either merge both or ensure this one only runs exactly when we need to publish a new image.

File: .github/workflows/docker_image.yml
Comment (sfan5) [inline .github/workflows/docker_image.yml:70]: how does this determine what to tag the image as?

the gitlab workflow [used to do](https://github.com/minetest/minetest/blob/master/.github/workflows/linux.yml#L155):
* `minetest/minetest/server:$CI_COMMIT_SHA`
* `minetest/minetest/server:$CI_COMMIT_REF_NAME`
* `minetest/minetest/server:latest`

Comment (AFCMS) [reply .github/workflows/docker_image.yml:70]: https://github.com/docker/metadata-action?tab=readme-ov-file#basic

Its different from the GitLab workflow by default, but it can be [customized](https://github.com/docker/metadata-action?tab=readme-ov-file#tags-input)

If required I can replicate the GitLab setup.

File: .github/workflows/docker_image.yml
Comment (AFCMS) [reply .github/workflows/docker_image.yml:15]: What could be done is an additional workflow job in this workflow where the image is pulled and then run `--version` inside it.

File: .github/workflows/docker_image.yml
Comment (AFCMS) [reply .github/workflows/docker_image.yml:9]: I don't know if there is a way to do this since we obviously want the image to be built on tags even if there aren't C++ changes.

One thing to be noticed is that the GitHub Actions cache feature is used here for caching Dockerfile steps, unlike the existing test in linux.yml, so a full rebuild will not happen everytime time and it will be much faster in case only Lua/Markdown are changed.

https://github.com/minetest/minetest/blob/e734b3f0d8055ff3ae710f3632726a711603bf84/.github/workflows/linux.yml#L155-L163

Comment (sfan5) [reply .github/workflows/docker_image.yml:9]: I suppose we could match against paths for pull requests, and build on every push.
(add a workflow condition not to tag PRs, obviously)

File: .github/workflows/docker_image.yml
Comment (sfan5) [reply .github/workflows/docker_image.yml:15]: can't that be added to this workflow? the image should be present after the buildx action runs

File: .github/workflows/docker_image.yml
Comment (sfan5) [inline .github/workflows/docker_image.yml:15]: since we don't publish no need for this branch filter, is there?

File: .github/workflows/docker_image.yml
Comment (sfan5) [inline .github/workflows/docker_image.yml:87]: cleaner IMO
(assuming this works, I didn't test it of course)
```suggestion
        run: docker run --rm $(cut -d, -f1 <<<"$IMAGE_TAGS") minetestserver --version
        shell: bash
        env:
          - IMAGE_TAGS: "${{ steps.meta.outputs.tags }}"
```

File: .github/workflows/docker_image.yml
Comment (sfan5) [inline .github/workflows/docker_image.yml:2]: (sorry for this nitpick but consistency)
```suggestion
name: docker_image
```

**Comment by sfan5** [comment]
Hmm, now it doesn't run for the PR anymore?
or is that because we just added it?

**Comment by AFCMS** [comment]
Failed to start: https://github.com/minetest/minetest/actions/runs/8254040414

**Review by sfan5 — APPROVED** [review]
thanks for working on this

**Comment by sfan5** [comment]
Consider also adding a note to README.md (like before 0b423dd061de9a0f7d5b80ea5a98dccd2eecec03).

**Comment by AFCMS** [comment]
I have tried adding back the kubernetes.yml file, but my IDE shows some errors

image

I never used Kubernetes and don't have the knowledge to fix those. I think I will just drop it (could be added back later if needed).

**Comment by AFCMS** [comment]
I still need to fix the image metadata

File: README.md
Comment (sfan5) [inline README.md:137]: i dont see any way how "latest" isn't equivalent to "master" except maybe for a short time after a release
so is this worth documenting?

File: README.md
Comment (sfan5) [inline README.md:144]: root is not generally needed for docker, depends on setup. so we shouldn't reproduce it in our instructions

File: README.md
Comment (sfan5) [inline README.md:170]: not sure about the syntax but this should be bound to localhost so people don't accidentally expose their prometheus

File: README.md
Comment (AFCMS) [reply README.md:137]: `latest` is the latest git tag, `master` is the latest build from the master branch

https://github.com/docker/metadata-action?tab=readme-ov-file#basic

Comment (sfan5) [reply README.md:137]: it's not clear at all that "latest" is that latest stable version

File: README.md
Comment (sfan5) [reply README.md:170]: yes

File: README.md
Comment (AFCMS) [reply README.md]: Better now?

File: README.md
Comment (sfan5) [inline README.md:137]:
```suggestion
* `ghcr.io/minetest/minetest:latest` (latest Git tag, which is the stable release)
```

File: README.md
Comment (sfan5) [inline README.md:141]:
```suggestion
See [here](https://github.com/minetest/minetest/pkgs/container/minetest) for all available tags.

For a quick test you can easily run:
```

File: README.md
Comment (sfan5) [inline README.md:148]:
```suggestion
To use it in a production environment, you should use volumes bound to the Docker host to persist data and modify the configuration:
```

File: README.md
Comment (sfan5) [inline README.md:169]: now nobody will be able to join the server
(30000/tcp = prometheus -> only for monitoring, internal; 30000/udp = Minetest server -> public)

**Review by sfan5 — CHANGES_REQUESTED** [review]

File: README.md
Comment (AFCMS) [reply README.md:168]: But isn't TCP used also for doing HTTP requests in mods/engine?

**Review by sfan5 — APPROVED** [review]

File: .github/workflows/docker_image.yml
Comment (rubenwardy) [inline .github/workflows/docker_image.yml:12]: What happens if you have both `branches` and `tags` in a workflow? Is it `and` or `or`?

File: README.md
Comment (rubenwardy) [inline README.md:131]: This should probably be moved to docs/docker_server.md or something, with a link here

**Comment by rubenwardy** [comment]
Looks good apart from that

File: .github/workflows/docker_image.yml
Comment (rubenwardy) [reply .github/workflows/docker_image.yml:12]: Ah, given that the docs mention latest master, it must be "or". Looks correct, resolving

**Review by rubenwardy — APPROVED** [review]
Looks good

**Comment by AFCMS** [comment]
On a side note after merge the package should be set as public in it's settings. At least on my account it's private by default and only visible to the repository maintainers (it's confusing IMO, there is no clear indication that it's private in the maintainers view).

**Pull request merged**
